src/stan/mcmc/hmc/xhmc/adapt_diag_e_xhmc.hpp

```diff
@@ -0,0 +1,55 @@
+#ifndef STAN_MCMC_HMC_xhmc_ADAPT_DIAG_E_XHMC_HPP
+#define STAN_MCMC_HMC_xhmc_ADAPT_DIAG_E_XHMC_HPP
+
+#include <stan/interface_callbacks/writer/base_writer.hpp>
+#include <stan/mcmc/stepsize_var_adapter.hpp>
+#include <stan/mcmc/hmc/xhmc/diag_e_xhmc.hpp>
+
+namespace stan {
+  namespace mcmc {
+    /**
+      * The No-U-Turn sampler (XHMC) with multinomial sampling
+      * with a Gaussian-Euclidean disintegration and adaptive
+      * diagonal metric and adaptive step size
+    */
+    template <class Model, class BaseRNG>
+    class adapt_diag_e_xhmc: public diag_e_xhmc<Model, BaseRNG>,
+                             public stepsize_var_adapter {
+    public:
+        adapt_diag_e_xhmc(Model &model, BaseRNG& rng)
+          : diag_e_xhmc<Model, BaseRNG>(model, rng),
+          stepsize_var_adapter(model.num_params_r()) {}
+
+      ~adapt_diag_e_xhmc() {}
+
+      sample transition(sample& init_sample,
+                        interface_callbacks::writer::base_writer& writer) {
+        sample s = diag_e_xhmc<Model, BaseRNG>::transition(init_sample,
+                                                           writer);
+
+        if (this->adapt_flag_) {
```

Comment:
This is so close to being duplicated code, only with variance instead of covariance plugged in.  But I don't see a good way to factor it without making it too hard to follow.